pkg/pillar/cmd/zedrouter/server.go

```diff
@@ -222,8 +249,290 @@ func (hdl hostnameHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		log.Errorf("No AppNetworkStatus for %s",
 			remoteIP.String())
 	} else {
+		w.Header().Set("Content-Type", "text/plain")
 		w.WriteHeader(http.StatusOK)
 		resp := []byte(anStatus.UUIDandVersion.UUID.String() + "\n")
 		w.Write(resp)
 	}
 }
+
+//processServiceState returns true if service is enabled
+func processMetadataService(ctx *zedrouterContext, w http.ResponseWriter, r *http.Request) bool {
+	log.Tracef("processMetadataService request: %s", r.URL.String())
+	if ctx.useMetadataService {
+		return true
+	}
+	http.Error(w, "Service disabled by configuration", http.StatusServiceUnavailable)
+	return false
+}
+
+// ServeHTTP for metadataHandler process requests for ec2-compatible requests
+func (hdl metadataHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !processMetadataService(hdl.ctx, w, r) {
+		return
+	}
+	dirname, filename := path.Split(strings.TrimSuffix(r.URL.Path, "/"))
+	dirname = strings.TrimSuffix(dirname, "/")
+	remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+	var hostname string
+	var id string
+	anStatus := lookupAppNetworkStatusByAppIP(hdl.ctx, remoteIP)
+	if anStatus != nil {
+		hostname = anStatus.DisplayName
+		id = anStatus.UUIDandVersion.UUID.String()
+	} else {
+		errorLine := fmt.Sprintf("no AppNetworkStatus for %s",
+			remoteIP.String())
+		log.Error(errorLine)
+		http.Error(w, errorLine, http.StatusServiceUnavailable)
+		return
+	}
+	switch filename {
+	case "meta-data":
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "instance-id")
+		fmt.Fprintln(w, "hostname")
+		fmt.Fprintln(w, "public-keys/")
+		return
+	case "hostname":
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, hostname)
+		return
+	case "public-keys":
+		anConfig := lookupAppNetworkConfig(hdl.ctx, anStatus.Key())
+		if anConfig == nil {
+			errorLine := fmt.Sprintf("no AppNetworkConfig for %s",
+				anStatus.Key())
+			log.Error(errorLine)
+			http.Error(w, errorLine, http.StatusServiceUnavailable)
+			return
+		}
+		keys, err := getSSHPublicKeys(hdl.ctx, anConfig)
+		if err != nil {
+			errorLine := fmt.Sprintf("cannot get ssh passwords for %s: %v",
+				anStatus.Key(), err)
+			log.Error(errorLine)
+			http.Error(w, errorLine, http.StatusInternalServerError)
+			return
+		}
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(http.StatusOK)
+		for ind := range keys {
+			fmt.Fprint(w, fmt.Sprintf("%d=key-%d", ind, ind))
```

Comment:
Shouldn't you be sending the keys here? 1=key-1 doesn't seem like useful info.

Reply:
It is few lines below: https://github.com/lf-edge/eve/pull/1951/files#diff-ef1656e32713a364dd39b3d84bf0ef0668a7872a311c460c7f35dafb659f6e37R321